src/axom/primal/geometry/NURBSCurve.hpp

```diff
@@ -0,0 +1,1177 @@
+
+    // Split the knot vector
+    KnotVectorType k1, k2;
+    m_knotvec.split(s, k1, k2);
+
+    n1.m_knotvec = k1;
+    n2.m_knotvec = k2;
+
+    // Copy the control points
+    n2.m_controlPoints.resize(nkts - s - 1);
+    n2.m_weights.resize(isCurveRational ? nkts - s - 1 : 0);
+    for(int i = 0; i < n2.m_controlPoints.size(); ++i)
+    {
+      n2.m_controlPoints[nkts - s - 2 - i] =
+        n1.m_controlPoints[n1.m_controlPoints.size() - 1 - i];
+
+      if(isCurveRational)
+      {
+        n2.m_weights[nkts - s - 2 - i] =
+          n1.m_weights[n1.m_weights.size() - 1 - i];
+      }
+    }
+
+    n1.m_controlPoints.resize(s - p + 1);
+    n1.m_weights.resize(isCurveRational ? s - p + 1 : 0);
+
+    if(normalize)
+    {
+      n1.normalize();
+      n2.normalize();
+    }
+  }
+
+  /// \brief Normalize the parameterization of the knot vector
+  void normalize() { m_knotvec.normalize(); }
+
+  /*!
+   * \brief Splits a NURBS curve (at each internal knot) into several Bezier curves
+   *   
+   * \return An array of Bezier curves
+   */
+  axom::Array<BezierCurve<T, NDIMS>> extractBezier() const
+  {
+    axom::Array<BezierCurve<T, NDIMS>> beziers;
+
+    const bool isCurveRational = this->isRational();
+
+    int p = getDegree();
+    if(p == 0)  // Handle this special case
+    {
+      for(int i = 0; i < getNumControlPoints(); ++i)
+      {
+        BezierCurve<T, NDIMS> bezier(0);
+        bezier[0] = m_controlPoints[i];
+
+        if(isCurveRational)
+        {
+          bezier.makeRational();
+          bezier.setWeight(0, m_weights[i]);
+        }
+
+        beziers.push_back(bezier);
+      }
+
+      return beziers;
+    }
+
+    // Split the curve at each knot value
+    int numBeziers = 1;
+    NURBSCurve<T, NDIMS> n1(*this);
+    for(int i = p + 1; i < m_knotvec.getNumKnots() - p - 1; ++i)
+    {
+      auto old_knot_count = n1.getNumKnots();
+      n1.insertKnot(m_knotvec[i], p);
+      auto new_knot_count = n1.getNumKnots();
+
+      if(new_knot_count != old_knot_count)
+      {
+        numBeziers++;
+      }
+    }
+
+    // For each Bezier, copy the control nodes into Bezier curves
+    BezierCurve<T, NDIMS> bezier(p);
+    if(isCurveRational)
+    {
+      bezier.makeRational();
+    }
+
+    int the_node = 0;
+    for(int i = 0; i < n1.getNumControlPoints(); ++i)
+    {
+      bezier[the_node] = n1[i];
+      if(isCurveRational)
+      {
+        bezier.setWeight(the_node, n1.getWeight(i));
+      }
+
+      the_node++;
+      if(the_node == (p + 1))
+      {
+        --i;  // Endpoint nodes are shared between Bezier curves
+        the_node = 0;
+        beziers.push_back(bezier);
+      }
+    }
+
+    return beziers;
+  }
+
+  /*!
+   * \brief Reset the degree and number of points in the curve
+   *
+   * \param [in] npts The target number of control points
+   * \param [in] degree The target degree
+   * 
+   * \note Will clear any data already in these arrays.
+   */
+  void setParameters(int npts, int degree)
```

Comment:
It might be worth noting that this will also make the knots uniform.

Is `setParameters` the best name for this function? (Full disclosure -- I didn't immediately think of a better name)

Reply:
I'm not too attached to `setParameters`. but I think it's more clear than `setSize` and more concise than `setNumControlPointsAndDegree`.